Terminal.Gui/Views/TreeView.cs

```diff
@@ -0,0 +1,1836 @@
+		/// Returns true if this branch has parents and it is the last node of it's parents branches (or last root of the tree)
+		/// </summary>
+		/// <returns></returns>
+		private bool IsLast()
+		{
+			if(Parent == null)
+				return this == tree.roots.Values.LastOrDefault();
+
+			return Parent.ChildBranches.Values.LastOrDefault() == this;
+		}
+
+		/// <summary>
+		/// Returns true if the given x offset on the branch line is the +/- symbol.  Returns false if not showing expansion symbols or leaf node etc
+		/// </summary>
+		/// <param name="driver"></param>
+		/// <param name="x"></param>
+		/// <returns></returns>
+		internal bool IsHitOnExpandableSymbol (ConsoleDriver driver, int x)
+		{
+			// if leaf node then we cannot expand
+			if(!CanExpand())
+				return false;
+
+
+			// if we could theoretically expand
+			if(!IsExpanded && tree.Style.ExpandableSymbol != null) {
+				return x == GetLinePrefix(driver).Count();
+			}
+
+			// if we could theoretically collapse
+			if(IsExpanded && tree.Style.CollapseableSymbol != null) {
+				return x == GetLinePrefix(driver).Count();
+			}
+
+			return false;
+		}
+
+		/// <summary>
+		/// Expands the current branch and all children branches
+		/// </summary>
+		internal void ExpandAll ()
+		{
+			Expand();
+
+			if(ChildBranches != null)
+				foreach (var child in ChildBranches) {
+					child.Value.ExpandAll();
+				}
+		}
+
+		/// <summary>
+		/// Collapses the current branch and all children branches (even though those branches are no longer visible they retain collapse/expansion state)
+		/// </summary>
+		internal void CollapseAll ()
+		{
+			Collapse();
+
+			if(ChildBranches != null)
+				foreach (var child in ChildBranches) {
+					child.Value.CollapseAll();
+				}
+		}
+	}
+
+	/// <summary>
+	/// Delegates of this type are used to fetch string representations of user's model objects
+	/// </summary>
+	/// <param name="model"></param>
+	/// <returns></returns>
+	public delegate string AspectGetterDelegate<T>(T model) where T:class;
+	
+	/// <summary>
+	/// Event arguments describing a change in selected object in a tree view
+	/// </summary>
+	public class SelectionChangedEventArgs<T> : EventArgs where T:class
+	{
+		/// <summary>
+		/// The view in which the change occurred
+		/// </summary>
+		public TreeView<T> Tree { get; }
+
+		/// <summary>
+		/// The previously selected value (can be null)
+		/// </summary>
+		public T OldValue { get; }
+
+		/// <summary>
+		/// The newly selected value in the <see cref="Tree"/> (can be null)
+		/// </summary>
+		public T NewValue { get; }
+
+		/// <summary>
+		/// Creates a new instance of event args describing a change of selection in <paramref name="tree"/>
+		/// </summary>
+		/// <param name="tree"></param>
+		/// <param name="oldValue"></param>
+		/// <param name="newValue"></param>
+		public SelectionChangedEventArgs(TreeView<T> tree, T oldValue, T newValue)
+		{
+			Tree = tree;
+			OldValue = oldValue;
+			NewValue = newValue;
+		}
+	}
+}
```

Comment:
You have my permission to break this class into multiple files. This is a lot of code to navigate in one file. 

Consider which if any of the classes might be generic and useful for purposes other than `TreeView`.

Reply:
I've split it up into `TreeBuilder.cs`, `TreeStyle.cs`, `TreeNode.cs` and `TreeView.cs`.  Currently the interfaces and implementations are in the same file.  Let me know if you want me to go further and put every class/interface in it's own file.

Currently `Branch<T>` is not public so I have left it in `TreeView.cs` but I can move it out too if required.